eth2util/types.go

```diff
@@ -6,12 +6,147 @@ import (
 	"encoding/json"
 	"strings"
 
+	eth2spec "github.com/attestantio/go-eth2-client/spec"
 	eth2p0 "github.com/attestantio/go-eth2-client/spec/phase0"
 	ssz "github.com/ferranbt/fastssz"
 
 	"github.com/obolnetwork/charon/app/errors"
 )
 
+// DataVersion defines the spec version of the data in a response.
+// The number values match those of go-eth2-client v0.17 and earlier releases.
+// This allows us to be compatible with those older versions when serialising as numbers.
+//
+// We should migrate to serialising as strings to aligned with eth2 spec at which
+// point this type can be removed in favour of the go-eth2-client type.
+type DataVersion string
+
+const (
+	DataVersionUnknown   DataVersion = ""
+	DataVersionPhase0    DataVersion = "phase0"
+	DataVersionAltair    DataVersion = "altair"
+	DataVersionBellatrix DataVersion = "bellatrix"
+	DataVersionCapella   DataVersion = "capella"
+	DataVersionDeneb     DataVersion = "deneb"
+)
+
+// legacyDataVersionValues maps DataVersion to the integer value used by go-eth2-client pre-v0.18.
+var legacyDataVersionValues = map[DataVersion]int{
+	DataVersionPhase0:    0,
+	DataVersionAltair:    1,
+	DataVersionBellatrix: 2,
+	DataVersionCapella:   3,
+	DataVersionDeneb:     4,
+}
+
+// MarshalJSON marshals the DataVersion as a number equaled to the go-eth2-client
+// pre-v0.18 integer value.
+func (v DataVersion) MarshalJSON() ([]byte, error) {
+	val, ok := legacyDataVersionValues[v]
+	if !ok {
+		return nil, errors.New("unknown data version")
+	}
+
+	b, err := json.Marshal(val)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal data version")
+	}
+
+	// TODO(corver): marshal as string in next step of migration in v0.18.
+
+	return b, nil
+}
+
+// UnmarshalJSON unmarshals the DataVersion from strings or a number equaled to the go-eth2-client
+// pre-v0.18 integer value.
+func (v *DataVersion) UnmarshalJSON(input []byte) error {
+	var strVal string
+	if err := json.Unmarshal(input, &strVal); err == nil {
+		if _, ok := legacyDataVersionValues[DataVersion(strVal)]; !ok {
+			return errors.New("unknown data version string")
+		}
+
+		*v = DataVersion(strVal)
+
+		return nil
+	}
+	var intVal int
+	if err := json.Unmarshal(input, &intVal); err != nil {
+		return errors.Wrap(err, "failed to unmarshal data version")
+	}
+
+	for version, val := range legacyDataVersionValues {
+		if intVal == val {
+			*v = version
+			return nil
+		}
+	}
+
+	return errors.New("unknown data version")
+}
+
+// ToUint64 returns the integer value used by go-eth2-client pre-v0.18.
+func (v DataVersion) ToUint64() uint64 {
+	return uint64(legacyDataVersionValues[v])
+}
+
+// ToETH2 returns a eth2spec.DataVersion equivalent to the DataVersion.
+func (v DataVersion) ToETH2() eth2spec.DataVersion {
```

Comment:
This method will return different values once we update go-eth2-client to v0.18, right?

Reply:
yup